technology-considerations/s3-user-access-to-objects.md

```diff
@@ -0,0 +1,121 @@
+* Using signed URLs means an eavesdropper, potentially, can grab the signed url and use it until it expires
+* AWS Cognito will work with AWS KMS encrypted AWS S3, can set up an AWS IAM or key policy allowing access to the AWS KMS key from AWS Cognito identities
+
+### Additional Considerations
+
+Unclear how AWS Cognito will work with Javascript silent login using Keycloak's Javascript library as the tokens are stored in Redis and not cookies.
+
+## Using AWS Cognito and Keycloak to Restrict Access to AWS S3 Bucket Objects
+
+### AWS IAM Policy To Restrict Access to AWS S3 Bucket Objects
+
+The following AWS IAM policy is an example of using an AWS Cognito identity sub parameter to control access to AWS S3 bucket objects.
+
+Note: The `cognito-identity.amazonaws.com:sub` refers to the identity id associated with the user in the AWS Cognito Identity Pool.
+
+```
+    {
+        "Version": "2012-10-17", 
+        "Statement": [
+            {
+                "Sid": "ListYourObjects",
+                "Effect": "Allow",
+                "Action": "s3:ListBucket",
+                "Resource": ["arn:aws:s3:::bucket-name"], 
+                "Condition": 
+                    {
+                        "StringLike": {
+                            "s3:prefix": ["cognito/application-name/${cognito-identity.amazonaws.com:sub}"]
+                        }
+                    }
+            }
+        ]
+    }
+```
+
+This policy restricts the listing of AWS S3 objects that contain a prefix including the identity id of the user.
+
+See the following for more details: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_s3_cognito-bucket.html
+
+### Using AWS IAM Policy With TDR
+
+TDR is using Keycloak to manage and authenticate users. 
+
+To be able to make use of an IAM policy outlined above to control access to AWS S3 objects, Keycloak needs to be integrated with AWS Cognito as an identity provider.
+
+This is because Keycloak users will need to be associated with an AWS Cognito identity, which can be used with the AWS IAM policy.
+
+Keycloak can be made an AWS Cognito authentication provider in two ways as:
+* a custom Developer authentication provider; or 
+* an OpenId authentication provider.
+
+Details of how to set up an AWS Cognito identity pool using these two approaches are available here:
+* Custom Developer Authentication provider: https://docs.aws.amazon.com/cognito/latest/developerguide/developer-authenticated-identities.html
+* OpenId Authentication provider: https://docs.aws.amazon.com/cognito/latest/developerguide/open-id.html
+
+There is additional detail on the authentication flow for both approaches available here: https://docs.aws.amazon.com/cognito/latest/developerguide/authentication-flow.html
+
+For the spike examples of using both methods were used. This was because the custom method was quicker to set up initially to help prove the IAM policy would restrict access to a user’s AWS S3 objects.
```

Comment:
Sorry if I've just missed it, but which method do you think we should use in production? I'd have guessed that the OpenID route is better because Keycloak _is_ an OIDC provider, so it's a standard workflow, but I haven't looked into the differences. 🙂

Reply:
Added explicit recommendation to use the OpenId method for production